### ISSUE TITLE
Remove squash from build

### ISSUE THREAD

**Issue by dnephin** [moby/moby#34565]
Now that we have multi-stage builds we don't need squash.

**Comment by cpuguy83** [comment]
I'm fine with removing it.
However unless there's some concrete reason to remove it (maintenance burden, preventing other features), I'd worry about upsetting people in the community.

**Comment by tonistiigi** [comment]
Although experimental feature means we can remove it anytime I think we could make an exception here and produce a warning for `17.09` and remove in `17.12`.

**Comment by dnephin** [comment]
Works for me. My primary reason for opening this issue is to document that we will likely be removing squash at some point. There was a request to add it to the Compose format, and I'd like to avoid adding it since it will be removed.

**Comment by vikstrous** [comment]
I don't agree that squash is not needed any more. You still can't produce an image with only one layer if you need to copy more than one file to more than one destination and run any commands.

**Comment by tonistiigi** [comment]
@vikstrous You can by just adding another stage from `scratch` and copying over the root from the previous stage. That being said, I don't think it is a good idea to promote usage like this. I've seen cases where people think that this will give them smaller images just because there are fewer layers but if the layers don't share files(like they don't for good multi-stage builds) there is no difference in speed if you use one layer on many. Only difference is that if you do squash it, you will not get build cache.

**Comment by vikstrous** [comment]
Ah, that sounds reasonable. It is an awkward workaround, but yeah, I see why it's an obscure use case.

**Comment by isanych** [comment]
Multi staged builds have nothing in common with squash. Multi stages builds could replace multiple Dockerfiles, squash replaces multiple commands. This code
```
FROM ubuntu

RUN apt-get update -qq \
 && apt-get install -y -qq squid3 \
 && apt-get clean \
 && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
 
COPY run.sh /
COPY squid.conf /etc/squid/
EXPOSE 3128
CMD ["/run.sh"]
```
Produces 5 images and it's harder to read and maintain because of tricks lowering number of images .
Below code looks better and easier to maintain and produces only 1 image with --squash
```
FROM ubuntu

RUN apt-get update -qq
RUN apt-get install -y -qq squid3
RUN apt-get clean
RUN rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
 
COPY run.sh /
COPY squid.conf /etc/squid/
EXPOSE 3128
CMD ["/run.sh"]
```

There are cases when multi staged build could replace squash - but it is a small subset of use cases for squash.

**Comment by cyberwillis** [comment]
I am working in my environment for data science that chains multiple dockerfiles until generate my final  one. The multi-stage way does not have any impact on it because there are many files, configurations bins, and dev-libs compiled in place that cannot simply be passed as an artifact.
Some experiments involve many packages together that I need build myself and I did used a modular approach (distributed dockerfiles), to keep simple to manage changes.
Squash had a huge impact on this, because dev libraries tend to escalate in each appended RUN command as @isanych mentioned. It's sad to see that will be gone soon.

**Comment by dnephin** [comment]
It sounds like this is some misunderstanding in this thread so I will attempt to clarify. Multi stage builds should handle every use case of squash, as well as many other use cases. Removing squash does not remove any features or functionality, because they are already provided by multi stage builds.

An image with 5 layers is not necessarily any larger or slower than an image with 1 layer. Extra layers are **only** a problem if you have to modify a bunch of files in multiple layers, which can be fixed by multi stage builds.

> There are cases when multi staged build could replace squash - but it is a small subset of use cases for squash.

@isanych I haven't found any cases where multi-stage build doesn't replace squash. Do you have any examples?

The first Dockerfile you posted is not a multi stage build. It uses best practices that have been established for years. Doing an `apt-get update` and `apt-get install` on different lines will not cache correctly and can cause the image build to fail when the index changes. The second example is wrong, and `--squash` does not fix it.  Building the first image is 185MB, which is the same size as the second with squash, so you don't even need multi-stage builds to get the equivalent sized image.

**Comment by isanych** [comment]
I think multi-stage could replace squash only in case when you produce artifacts in single known location.
```
FROM some_image
ADD HugeArchive.tar /tmp
RUN step1
...
RUN step99
RUN make install # copy lot of files all over the place
RUN rm -rf /tmp/*
```
I could not use multi stage as I don't know where files installed (yes, I could discover all of them and maintain up to date list of files after changes - but it unnecessary and error prone). 

Without squash I have to use workaround:
a) put HugeArchive.tar on web server
b) use very long RUN command so all temporary files will not be a part of final image
As a result I cannot use docker cache as all 99 steps are part of the same run command so every time I change step 99 all previous steps have to be executed.

so without squash most of my Dockerfiles looks like:
```
FROM some_image
COPY run.sh /tmp
RUN /tmp/run.sh && rm -rf /tmp/*
```
In my use cases I cannot benefit from multi-stage builds, in cases when I produce artifacts in known location I upload them into artifact repository and use them when they needed without building them first.

On windows squash is not working correctly (I only recently discovered it and not created an issue as squash is intended to be removed anyway), so I have to use old approach with single script file and it is less convenient than multiple RUN commands and squash.

I definitely could survive without squash using workarounds, and if maintaining it is a big burden I could understand that argument, but I don't agree that multi stage builds make squash redundant.

**Comment by cpuguy83** [comment]
@isanych What's the purpose of creating artifacts at unknown locations?

**Comment by isanych** [comment]
@cpuguy83 most of make install commands create multiple files in multiple locations (binaries, configs, mans, etc). I know that I should install packages from repo, but I could not use repos in rhel5 and even when it possible to use centos5 most of stuff is outdated there (I know that it past EOL and official images don't works with repos out of the box, my images are working). But rhel5 is the only linux which we officially support (I even had centos3 image - luckily I did not needed it for a while). So keep it in mind please that some users still use 32 bit and/or ancient images :)

**Comment by dnephin** [comment]
> most of make install commands create multiple files in multiple locations

I believe most `make install` commands also allow you specify a "base path" as well, right? Which might default to `/usr`, but you could also specify a different directory (that is not used by anything else), which would make it very easy to copy the results from a single directory tree into a following stage.

**Comment by isanych** [comment]
Yes, when I use `make install --prefix` I have all files in one location, and it that case I achive that location and upload to aftifact repository - I don't use multi stage builds in that case, but `make install` & `make install --prefix` produce different results. As I said before I'm capable to find workarounds for all my use cases and I was using docker long before squash option was introduced.
The only point which I'm trying to make that removal of squash will force me to use workarounds and at least at the moment I don't see single case where I could use multi-stage builds in these workarounds. And workarounds are still workarounds - straight road is almost always better imho.

**Comment by ijc** [comment]
Am I confused/missing something or can any `Dockerfile` not be made to behave as if `--squash` was used by simply appending two lines:
```
FROM scratch
COPY --from=0 / /
```
?

**Comment by isanych** [comment]
I don't think so. I will be equivalent of docker save - you loose base container information and all your images will have all files instead of using files from base image.

If you could do:
```
FROM ubuntu
...
FROM ubuntu
COPY --from=0 / / # only changed files copied here
```
then it will be equivalent of squash

**Comment by ijc** [comment]
So given `LAYER 0 (base), LAYER 1, LAYER 2, ... LAYER N` then `--squash` will produce `LAYER 0 (base), LAYER 1toN` as opposed to producing only `LAYER 0toN`? I wasn't aware of that, and indeed if that is the case it sounds like they aren't directly equivalent then.

**Comment by thaJeztah** [comment]
Correct; reasoning behind that was indeed to still allow sharing base images, but only squashing what was _added_ to the base image. I know this was brought up at some point during the implementation of `COPY --from` (but couldn't find where it was discussed); IIRC, it was left for a future enhancement to have something like `--from ..  --skip-base` (or whatever syntax would work).

Note that for combining multiple `RUN` statements, without having to chain all commands together, I also opened https://github.com/moby/moby/issues/34423 ("Proposal: add "heredoc" notation to the Dockerfile syntax"). There's also a proposal for `EXPORT` / `IMPORT` in https://github.com/moby/moby/issues/32100, which may be relevant, but there's some concerns raised about it being too complicated.

**Comment by thaJeztah** [comment]
Opened https://github.com/moby/moby/issues/35280 as a proposal to address the "squash" use case

**Comment by tonistiigi** [comment]
Left a comment in #35280 how duplicates detection is driver dependant atm and already works correctly in a lot of cases. Another way to have even more control is with https://github.com/moby/moby/issues/32507 that would let you apply differences of 2 arbitrary directories(2 points from your dockerfile).

The concept of a layer is arbitrary. In a long view, the user shouldn't care how many layers image has or if it has any layers at all or uses some other mechanism for distribution, or if certain Dockerfile command creates a layer or not. They should just know that their images don't contain unnecessary data and the push/pull commands are efficient.

**Comment by duglin** [comment]
Just to add some additional info.... when I do Docker education classes I go out of my way to talk about the optimizations that Docker does to help speed things up and to make better use of resources. And the layering within images is part of that. 

While I agree that in many ways layers are an internal artifact of Docker images, I'm not sure I would say that they should be totally hidden from users in all cases because people do/will make decisions based on them. For example, knowing that if the same base image is used it will result in faster downloads (because they just pull down the extra additive layers for subsequent containers), and that same base layer will be reused (instead of duplicated) for each container is a really nice optimization that people like. And in order to leverage that they need to think ahead about how they structure their layers/images - even if its just to use the same base image.

**Comment by tonistiigi** [comment]
> For example, knowing that if the same base image is used it will result in faster downloads (because they just pull down the extra additive layers for subsequent containers), and that same base layer will be reused (instead of duplicated) for each container is a really nice optimization that people like.

I agree on that and that's what I meant by "push/pull commands are efficient". Users should know that using same base image makes resulting images more efficient but ideally, they shouldn't need to care about how this is achieved on the wire. For example, if images used a rootfs that is a DAG tree of files(or a continuity manifest), this property would still apply, but there wouldn't be any concept of layers.

**Comment by galuszkak** [comment]
I don't know if I can post a question here, but for me there is one thing that is maybe not covered by multistage builds (or I have to small knowledge about this feature).

We have Dockerfile that add `pip.conf` file with sensitive credentials, then we install dependencies and as a next step we remove that file. After that we squash image, so anyone who will use that image won't have access to layer with that credentials.

Now per my understanding, there is no way to copy layer with final resolution of install dependencies in docker multi-stage installation. Is there easy way to say I want You to copy particular final layer to next image?

**Comment by thaJeztah** [comment]
@galuszkak with multi-stage builds, you'd do something like this:

```Dockerfile
FROM busybox AS stage-one
COPY ./my-secret.thing ./somewhere/my-secret.thing
RUN echo "build some things"
RUN rm ./somewhere/my-secret.thing

FROM busybox
RUN env
COPY --from=stage-one /the-things-i-built /app/foo/
```

The "secrets" are kept in the first stage, and (in the example above) removed before you copy the build-artifacts to the second / final stage.

If you don't `"docker push"` the image of the intermediate stage (only the image you produced by the final stage), those credentials /secrets are not distributed.

Be aware though that the secrets _will be_ in the local image cache (which is where the image built by the first stage is stored). This should not be a problem if the server on which docker is running is trusted, but keep in mind that anyone that has access to that daemon / host can access the image cache (so access those credentials).

**Comment by dhirschfeld** [comment]
> Multi stage builds should handle every use case of squash, as well as many other use cases. Removing squash does not remove any features or functionality, because they are already provided by multi stage builds.

@dnephin I gave your comment a 👎 above because this is simply not the case for Windows containers. Many windows installers make changes to the environment and registry which would be near impossible to reverse-engineer into a new image - you have run the installer in the image where it is to be used. `--squash` allows the builder to remove unnecessary bloat from complicated installs which isn't possible to do with just multi-stage builds. Unfortunately `--squash` has never worked on Windows due to #31468.

Multi-stage builds and `--squash` are two different ways of accomplishing the same thing, but both work in different situations and both are useful.

**Comment by duglin** [comment]
And not just for windows. Even in a Linux environment if some installer did put things in only certain spots, from a UX perspective asking the person setting up the Dockerfile to track all of them down (and monitor any potential changes in that install process as new versions comes out) is a lot to ask.

IMO mutli-stage builds, squash and INCLUDE are different things and while they do have some overlapping usecases, they also have non overlapping ones.

**Comment by cyberwillis** [comment]
In my case also I use docker as environment not as server for some isolated app, so I need to build many applications and install it locally and so files and configurations spread on the system, being   almost untraceable about what changed what. The Multi-stage never worked in my setup and never will because it does overwrite files in the process. My container has around 10 Gb of size and I use it like a Node Machine not as a web server like multi-stage are inclined to contribute at most of cases.

**Comment by alexivkin** [comment]
For fox sake, just fix --squash on windows. Windows apps modify so much internal stuff that keeping track of it all and then copying to the next stage is nearly impossible.

**Comment by dhirschfeld** [comment]
@alexivkin - while I can appreciate the sentiment I think it's best to keep the discussion on a professional level. Thinly veiled profanity is best left for vendors whom we actually pay for their product - I don't think it will advance our cause in this instance.

**Comment by dnephin** [comment]
> Wouldn't you have to redefine ENV and other metadata multiple times if you were forced to use multi-stage builds?

Only if different stages used the same ENV variables, which shouldn't be the common case if stages are used correctly for independent stages of the build.

This would also be fixed by #32100 allowing you to export ENV from the build stage to the runtime stage.

**Comment by veqryn** [comment]
Well, it would be more common if you are forced to use multi-stage builds JUST to squash your images.

For example:
```dockerfile
FROM oracle AS needs-squashing
ENV NEEDED_VAR some_value
COPY ./giant.zip ./somewhere/giant.zip
RUN echo "install giant in zip"
RUN rm ./somewhere/giant.zip

FROM scratch
COPY --from=needs-squashing / /
ENV NEEDED_VAR some_value
```

I still like `squash` as a simple way to handle one of the most common use cases, even if what it does can technically also be done with more complicated syntax.

**Comment by dhirschfeld** [comment]
If you can do that in general after installing a Windows MSI I'll eat my :tophat:!

**Comment by thaJeztah** [comment]
> Lets say both an install script and the final product at runtime needed an ENV var set. If we don't have squash, we'd have to define that multiple times.

Here's a little trick that would work for that:

```Dockerfile
# Set ARG outside of first build context, and assign a default value
ARG I_AM_EVERYWHERE=woot

FROM busybox AS stage1

# No default value set, so the value is set to the default that's set at the top
ARG I_AM_EVERYWHERE
RUN echo inside stage 1 $I_AM_EVERYWHERE


FROM busybox AS stage2
ARG I_AM_EVERYWHERE
RUN echo inside stage 2 $I_AM_EVERYWHERE

# And if you want to "bake" the value into the image
ENV I_AM_EVERYWHERE=$I_AM_EVERYWHERE
```

Building that should show:

```
Step 4/8 : RUN echo inside stage 1 $I_AM_GLOBAL
 ---> Running in 0e55c8f23b25
inside stage 1 woot

...


Step 7/8 : RUN echo inside stage 2 $I_AM_GLOBAL
 ---> Running in 268b3486e460
inside stage 2 woot
```

**Comment by kiorky** [comment]
|``COPY --from`` is way far from squashing as **it doesnt preserve permissions**

**Comment by isanych** [comment]
Maybe we could introduce SQUASH command (with optional parent parameter (including no parent - like docker save), default parent is latest FROM image)?

This way we would not need to propagate --squash option to different tools and squash command could be cached (currently squash happens every time, even when all build steps are cached).

**Comment by kiorky** [comment]
Yep, i'm also +1 to enhance squashing up to dockerfile, more over with the ability to specify the ancestor to squash from (today we can only squash from direct ancestor with builtin ``--squash``).

**Comment by jsirex** [comment]
I believe we shouldn't remove `--squash`:
1. If you don't want it - DON'T USE it then!
2. Sometimes `--squash` is **only** way to solve really typical issue
3. Building docker images without understanding *layer* produces many intermediate layers, which makes serious impact on images. That's why, for "now" you cannot hide *layers* implementation from user. That's why user takes care about this and writes ugly "cmd1 && cmd2 && cmd3"

Really typical use-case:
1. build project which produces packages (make deb)
2. create container based on deb from step 1.

```Dockerfile
FROM any-image
ADD ./my-project.deb
RUN apt-get install ./my-project.deb
RUN rm ./my-project.deb
```

I'm sure that removing intermediate package **IS NOT POSSIBLE** without squashing.
1. Multi-Staged builds won't help.
2. You, probably, can upload package to a http server or run it somehow locally during docker build, but it looks like ugly hack.

If think we can remove this only after something new will have been introduced to solve this particular scenario.

**Comment by cpuguy83** [comment]
@jsirex 
```Dockerfile
RUN --mount source=./my-project.deb,target=./my-project.deb  dpkg -i ./my-project.deb
```

This is an open proposal for mounting from either
1. an image
2. the actual build context

This is mounting instead of copying, which means there is nothing that needs to be removed.

**Comment by jsirex** [comment]
@cpuguy83 
What if I need job done right now? Currently I have `--squash`.
Depending on dockers' speed when `--mount` will come out? in 2020? 2025? 2073 :) ?
I'm just pointing that this feature must not be removed until another has implemented.

PS. Still looks like hack. Probably, docker should mount whole build context somewhere... IDK

**Comment by dhirschfeld** [comment]
@cpuguy83 - also does nothing for the "unknown artifacts" problem which squash is the only feasible way to solve.

**Comment by staticfloat** [comment]
I also would like to keep squashing: I use this in a rather [large set of dockerfiles](https://github.com/staticfloat/julia-docker/tree/master/workerbase) that have upwards of 200 layers commands in a single image; it's really helpful to have the intermediate build caching that happens when using Docker, but many of these build images have tens of gigabytes of temporary files generated within them, (we build cross-compilers for many different architectures; imagine extracting, compiling, and then deleting the entire GCC source tree a dozen times, once for each target architecture) and `docker build --squash` reduces the final image size to something like 4GB instead of 40GB.  This could, of course, be done through chaining lots of commands together into a single `RUN` statement, but then we lose a lot of the expressive power of Dockerfiles, and this seems like a poor substitute for the `--squash` command.

**Comment by cpuguy83** [comment]
@staticfloat sounds like you should  eat using multistage builds.

**Comment by varhub** [comment]
Im agree with @isanych , @dhirschfeld  @cyberwillis , ..., and guess that @cpuguy83 is just making a stupid naive simplification of real world.
To end up with *really* what multistage provides and why squash is another thing not-covered by multistage I propose this simple exercise to @cpuguy83 :
Tell us one usage case of multi stage that cannot be covered by:
* run all commands on a container
* copy created files to a `-v` mount point
* copy command on new brand Dockerfile from context

I known, this is annoying and multi stage made a huge job simplifiying this use case.
But rationale reduces to assuming a blackbox or whitebox approach.

PS: notice that make --prefix relies on perfectly well job done by developers. And almost all cases of cmake that do not end into a .dkpg just hardcore absolute paths

**Comment by cpuguy83** [comment]
@varhub I'm not sure I understand this flow. Can you give a basic example of this?

**Comment by varhub** [comment]
@cpuguy83 I will try to explain it.
Suppose that you have an already created container with all build-essential requirements. With docker run, or even run -it, execute all commands (of dockerfile RUN).
Then, copy all files to host by running docker run --volume.
Finally, a simple dockerfile that just copy these files from host to container.
Do you feel the similarity with multistage?
This is more or less the workflow that I was refering.

**Comment by thaJeztah** [comment]
But why copy them from the container to the host, only to copy them back during build?

note: If this is a pre-built image, you can also copy from that image in a build step;

```Dockerfile
COPY --from=myimage:latest /foo /bar
```

**Comment by kiorky** [comment]
> It looks like it looses file ownerships. For my case, this means sudo is borked.

+ it would not act like a ``rsync -a --delete`` (wont remove files that are already in B) and it's problematic if you won't start from scratch (not the case here).

**Comment by tonistiigi** [comment]
> **it doesnt preserve permissions (and ownerships)**
> It looks like it looses file ownerships. For my case, this means `sudo` is borked.

This has been fixed in buildkit and being backported to legacy in https://github.com/moby/moby/pull/38599

**Comment by agowa338** [comment]
I want to request to add the ability for Copy and Add to not generate a new layer.
So one could use multistage build to build a application in step 1 and than install it within stepp 2.
Currently if one does that the application gets duplicated, as the copy instruction as well as the following RUN instruction generates a new layer.

If we could do something like:
```
FROM alpine:latest as build
RUN apk --no-update add git ...
RUN git clone ...
RUN make

FROM alpine:latest
COPY --from-stage=build /project/package.apk /tmp
RUN apk add --allow-untrusted /tmp/package.apk && rm /tmp/package.apk
```
The rm can as well be removed, as the COPY instruction generates a new layer.
If there was a `--no-layer` parameter to `COPY` and `ADD` that would also resolve this issue for me.

If you think that having that package within that layer is no problem, than imagine it being chromium, wine, powershell, LaTeX, or anything else that has very big packages....

**Comment by lephuongbg** [comment]
I hope that squash won't be removed unless COPY (or may be a new command) can reliably only copy changed files between stages: https://github.com/moby/moby/issues/21950

We are now having a build procedure that looks like this:

```Dockerfile
FROM base-image as build
RUN /build/build-a-multi-gigabytes-project-folder.sh

# Use current production image as base layer
# then add only changed files on top of it for faster pull on production server
FROM current-production-image
COPY --from=build /build /build
RUN rsync -rlpgoDc --stats /build /production/
# Remove the build folder and utilize squash to completely remove it from final image
RUN rm -r /build
```

**Comment by thaJeztah** [comment]
When using BuildKit (`DOCKER_BUILDKIT=1`) and the [experimental Dockerfile syntax](https://github.com/moby/buildkit/blob/b521aae3ea1e1bb73298481ac538f9880ccd3854/frontend/dockerfile/docs/experimental.md), you could use `RUN --mount` to mount the files from the first stage, without them being copied into the second stage; so something like:


```Dockerfile
# syntax=docker/dockerfile:experimental 
FROM base-image as build
RUN /build/build-a-multi-gigabytes-project-folder.sh

# Use current production image as base layer
# then add only changed files on top of it for faster pull on production server
FROM current-production-image
RUN --mount=type=bind,from=build,src=/build,target=/build \
  rsync -rlpgoDc --stats /build /production/
```

**Comment by garo** [comment]
We have a dozen or so projects which all start "FROM ubuntu" and then they build upon that. --squash allows us to flatten the project changes into a single layer while still inheriting from the base image, which saves both space and speeds up deployments as the base image can be shared.

Using multistage builds doesn't allow this as it would also squash the base image in.

**Comment by Heatmanofurioso** [comment]
+1

**Comment by paralin** [comment]
If there was some way to mount a previous stage of the multi-stage build, you could use "rsync"

FROM mybaseimage
MOUNTPREV mystage /my-stage
RUN rsync -rv --delete /my-stage /

etc

Still messy.

**Comment by alexschoeberl** [comment]
Instead of an argument for build, a squash option could be added to FROM.

    FROM base as prepare
    RUN apt install ...
    RUN apt install ...
    
    FROM prepare --squashed

In this case only the base image and the other layers squashed together would be kept.

**Comment by riverar** [comment]
Docker cannot copy `C:\` or `C:\Windows` so what's the proposed alternative for Windows containers?

**Comment by thaJeztah** [comment]
@lowenna do you recall if there was a technical reason to disallow copying c:\ and c:\windows? I see it was added in https://github.com/moby/moby/pull/34252, but I couldn't find a mention if there's a specific reason

**Comment by lowenna** [comment]
@thaJeztah I don't recall a specific reason apart from it being a pretty-much nonsense operation to attempt. That said, it was almost certainly explicitly blocked as someone tried, and for a reason I cannot recall, it failed. FWIW though, `squash` never worked on Windows (WCOW) due to underlying platform limitations

**Comment by IBMRob** [comment]
A prime use case we are using this for is when installing system packages using a package manager. Using multi-stage with these is very hard as its not just a case of saying "Its installed files in X, Y and Z, copy them into the next stage". A simple example of what we are doing is
```dockerfile
FROM BASEIMAGE
COPY largeTar.gz /tmp
RUN tar -xvf /tmp/*.tgz && \
  rpm -ivh /tmp/*  && \
  rm -rf /tmp
 ```
If we use squash in the example above we have installed the package properly, using the proper package manager but it only adds the extra files, the install binaries are not included.

**Comment by paralin** [comment]
This trick works fine in all cases I've tried:

```Dockerfile
FROM ubuntu:focal as stage1

# setup environment
ENV container=docker \
  LANG=en_US.UTF-8 \
  LANGUAGE=en_US:en \
  LC_ALL=en_US.UTF-8

# packages
RUN export DEBIAN_FRONTEND=noninteractive; \
  apt-get update && \
  apt-get dist-upgrade -y && \
  apt-get install -y  \
  --no-install-recommends \
  -o "Dpkg::Options::=--force-confdef"  \
  -o "Dpkg::Options::=--force-confold"  \
  build-essential \
  openssh-client \
  rsync \
  lsb-release \
  systemd \
  usbutils \
  sudo \
  curl \
  htop \
  git \
  vim \
  nano \
  net-tools \
  software-properties-common \
  unzip \
  autotools-dev \
  cups \
  lightdm \
  xorg \
  xserver-xorg-input-all \
  locales \
  ubuntu-desktop-minimal \
  wget && \
  apt-get autoremove -y && \
  rm -rf /var/lib/apt/lists/*

RUN locale-gen "en_US.UTF-8"

FROM scratch as stage2

COPY --from=stage1 / /

ENV container=docker \
  LANG=en_US.UTF-8 \
  LANGUAGE=en_US:en \
  LC_ALL=en_US.UTF-8

```

**Comment by thaJeztah** [comment]
> This trick works fine in all cases I've tried:

The downside of that approach is that _everything_ is squashed to a single layer, including the Ubuntu base image. Generally that's something you'd like to avoid, because it means that that base image will no longer be shared with other images also based on the same Ubuntu base image.

A better approach would be to use `FROM ubuntu:focal` for the final stage, but unfortunately depending on the storage driver, that's not always an option currently (see https://github.com/moby/moby/issues/35280)

_However_, in your particular example, I'd actually recommend _not_ squashing at all; the `ENV` instructions only modify metadata of the image (by themselves don't add size, or layers), and the `RUN` step only _adds_ files (and those files don't overlap with existing files in the base image). Keeping those layers separate won't make the image larger, and likely will make it _faster_ to pull the image, because the layers of the Ubuntu base image may already be in the local image cache; allowing docker to only pull the layers that were added "on top".

If your Dockerfile would also have a step that (e.g.) does a `COPY` of a local file

```dockerfile
RUN locale-gen "en_US.UTF-8"
COPY my-config.ini /etc/config/my-config.ini
```

Making a change to those file(s), even if the files are just "small" (the config example above), (with your "squash" approach) would produce a _completely new_ layer with everything. (So, e.g. `1 byte changed in my-config.ini` + `ubuntu:focal (~80MB)` + `RUN apt-get .... (xxxMB)`).


Having multiple layers _in itself_ isn't bad (and for reasons above, actually _good_). That is, of course "within limits" (applying 80 layers may take longer when first running an image, and there are limits to the number of layers).
Squashing layers is mostly useful if multiple steps touch the same files, or if subsequent steps in the Dockerfile _remove_ files from earlier steps.

**Comment by agowa338** [comment]
I'm using BuildKit nowadays for that.
Just use `docker buildx build` instead of `docker build` (or alternatively run `export DOCKER_BUILDKIT=1` before `docker build`).
Than you can do:
```
RUN --mount=type=cache,target=/var/cache/apt # Mount a cache directory
RUN --mount=type=ssh # Use hosts ssh keys
RUN --mount=type=secret,id=yoursecret # mount secret in /run/secrets/yoursecret
RUN --mount=type=bind,from=buildContainer,source=/,target=/mnt dpkg -i /mnt/path/to/build.deb # Mount source folder from build container temporarily into the main container and execute the installation in one step.
```

**Comment by GrmanRodriguez** [comment]
> I've seen cases where people think that this will give them smaller images just because there are fewer layers but if the layers don't share files(like they don't for good multi-stage builds) there is no difference in speed if you use one layer on many. 

Hi. Jumping it to mention that I recently opened an issue where this behavior is not happening #43614 @tonistiigi

**Comment by jkt628** [comment]
i have a case where i remove the JRE from a vendor supplied image and install a different JRE.  it would be nice to [specify exactly which layer gets squashed](https://github.com/moby/moby/issues/34565#issuecomment-718698020) to preserve everything EXCEPT the original JRE and then put my layers, including the different JRE, on top.

**Comment by cyrush** [comment]
Now that squash appears to be gone:
```
WARNING: experimental flag squash is removed with BuildKit. You should squash inside build using a multi-stage Dockerfile for efficiency.
```

Here is an oversimplified use case that I would like to understand if multi-stage builds can help with:

```
RUN command-that-creates-a-large-set-of-temporary-intermediate-files-as-a-side-effect
RUN command-that-cleans-up-temporary-intermediate-files
```

With squash, wouldn't those intermediate files not appear in the final image and lead to reduction in overall size vs a full set of layers? 

Can a multi stage build help accomplish this type of reduction - or is the only way to chain those commands into a single `RUN`?

Thanks to anyone with pointers for a use case like this!

**Comment by agowa** [comment]
@cyrush yes and no. As I said two posts earlier you can do:

> I'm using BuildKit nowadays for that. Just use `docker buildx build` instead of `docker build` (or alternatively run `export DOCKER_BUILDKIT=1` before `docker build`). Than you can do:
> 
> ```
> RUN --mount=type=cache,target=/var/cache/apt # Mount a cache directory
> RUN --mount=type=ssh # Use hosts ssh keys
> RUN --mount=type=secret,id=yoursecret # mount secret in /run/secrets/yoursecret
> RUN --mount=type=bind,from=buildContainer,source=/,target=/mnt dpkg -i /mnt/path/to/build.deb # Mount source folder from build container temporarily into the main container and execute the installation in one step.
> ```

In your case the `command-that-creates-a-large-set-of-temporary-intermediate-files-as-a-side-effect` would have to be in one stage and the `command-that-cleans-up-temporary-intermediate-files` in another while mounting directories.
OR the good old `RUN command-that-creates-a-large-set-of-temporary-intermediate-files-as-a-side-effect && command-that-cleans-up-temporary-intermediate-files`.

**Comment by cyrush** [comment]
@agowa thank you, for simplicity we will used chained commands -- but good to know about other ways to think about this case.
